src/main/java/ladder/domain/ladder/Ladder.java

```diff
@@ -1,38 +1,72 @@
 package ladder.domain.ladder;
 
+import ladder.domain.ladder.shape.LadderShapeInfo;
 import ladder.exception.ErrorMessage;
 
+import java.util.Collections;
 import java.util.List;
 import java.util.Objects;
 import java.util.stream.Collectors;
+import java.util.stream.IntStream;
 import java.util.stream.Stream;
 
 public class Ladder {
 
-    private final List<Pillar> pillars;
+    private static final int FIRST_INDEX = 0;
 
-    private Ladder(final List<Pillar> pillars) {
-        validatePillars(pillars);
-        this.pillars = pillars;
+    private final List<RowPillars> rowPillars;
+
+    private Ladder(final List<RowPillars> rowPillars) {
+        validatePillars(rowPillars);
+        this.rowPillars = rowPillars;
+    }
+
+    private void validatePillars(final List<RowPillars> rowPillars) {
+        if (Objects.isNull(rowPillars) || rowPillars.size() < Height.MIN_HEIGHT) {
+            throw new IllegalArgumentException(ErrorMessage.IS_NULL_OR_EMPTY);
+        }
     }
 
-    public static Ladder of(final Height height, final int numOfPlayers) {
-        return new Ladder(createPillars(height, numOfPlayers));
+    public static Ladder of(final LadderShapeInfo ladderShapeInfo) {
+        return new Ladder(createPillars(ladderShapeInfo));
     }
 
-    private static List<Pillar> createPillars(final Height height, final int numOfPlayers) {
-        return Stream.generate(() -> Pillar.of(PillarCount.of(numOfPlayers)))
-                .limit(height.getValue())
+    private static List<RowPillars> createPillars(final LadderShapeInfo ladderShapeInfo) {
+        return Stream.generate(() -> RowPillars.of(ladderShapeInfo.getWidth()))
+                .limit(ladderShapeInfo.getHeight())
                 .collect(Collectors.toList());
     }
 
-    private void validatePillars(final List<Pillar> pillars) {
-        if (Objects.isNull(pillars) || pillars.size() == 0) {
-            throw new IllegalArgumentException(ErrorMessage.IS_NULL_OR_EMPTY);
+    public List<RowPillars> getRowPillars() {
+        return rowPillars;
+    }
+
+    public List<Integer> ride() {
+        if (isSinglePlayer()) {
+            return Collections.singletonList(FIRST_INDEX);
         }
+
+        return IntStream.range(FIRST_INDEX, getLadderWidth())
+                .map(this::rideOnePlayer)
+                .boxed()
+                .collect(Collectors.toList());
+    }
+
+    private int rideOnePlayer(final int playerPosition) {
+        int position = playerPosition;
+
+        for (RowPillars pillar : rowPillars) {
+            position = pillar.move(position);
+        }
+
+        return position;
```

Comment:
```suggestion
        return rowPillars.stream()
                .reduce(playerPosition, (byPosition, rowPillar) -> rowPillar.move(byPosition),
                        (x, y) -> {throw new RuntimeException("Can't reduce in parallel environment");});
```
이런식으로 reduce를 사용할 수 있어요! 세번째 파라미터는 병렬처리를 막기위해서 예외를 던지도록 했습니다.

Reply:
https://www.baeldung.com/java-stream-reduce
https://www.baeldung.com/java-stream-sum

Reply:
구현한 사다리 타기 게임에서 각 행(RowPillars) 사이에는 순서 보장이 필요한 구조
- 스트림의 경우 순서에 상관 없이 병렬 처리가 가능
- 세번 째 파라미터는 병렬 처리 결과 간의 값들을 어떻게 처리할지에 대한 방법 정의
- 병렬 처리로 수행하는 경우, 예외 발생 (stream 대신 parallelStream 을 사용하는 경우)